### PR TITLE
[13.0][ADD] sale_payment_sheet: New module

### PR DESCRIPTION
**Pull request by sergio-teruel**
New module that allows to create payments to commercial users.
cc @Tecnativa TT27097
ping @carlosdauden @victoralmau

### CODE REVIEW

File: sale_payment_sheet/models/sale_payment_sheet.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+        for sheet in self:
+            if sheet.state != "open":
+                raise UserError(
+                    _("You can not delete a sheet if has related journal items.")
+                )
+        return super().unlink()
+
+    def button_confirm_sheet(self):
+        sheets = self.filtered(lambda r: r.state == "open")
+        BankStatement = self.env["account.bank.statement"].sudo()
```
Comment (victoralmau) [inline sale_payment_sheet/models/sale_payment_sheet.py:133]:
```suggestion
        BankStatement = self.env["account.bank.statement"].sudo()
        BankStatementLine = self.env["account.bank.statement.line"].sudo()
```

File: sale_payment_sheet/models/sale_payment_sheet.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+                    "date": sheet.date,
+                    "journal_id": sheet.journal_id.id,
+                    "user_id": sheet.user_id.id,
+                }
+            )
+            for line in sheet.line_ids:
+                statement_line = (
+                    self.env["account.bank.statement.line"]
+                    .sudo()
+                    .create(
```
Comment (victoralmau) [inline sale_payment_sheet/models/sale_payment_sheet.py]:
```suggestion
                statement_line = BankStatementLine.create(
```

**Review by victoralmau — APPROVED** [review]
Good job!

**Comment by sergio-teruel** [comment]
@victoralmau Changes done!!

File: sale_payment_sheet/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+# Copyright 2020 Tecnativa - Sergio Teruel
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+{
+    "name": "Sale payment sheet",
+    "summary": "Allow to create invoice payments to commercial users without "
+    "accounting permissions",
+    "version": "13.0.1.0.0",
+    "development_status": "Beta",
+    "category": "Account",
+    "website": "https://gitlab.tecnativa.com/Tecnativa/marmenorda-odoo",
```
Comment (rousseldenis) [inline sale_payment_sheet/__manifest__.py]: Please change this

File: sale_payment_sheet/models/sale_payment_sheet.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+from odoo.exceptions import UserError, ValidationError
+from odoo.tools import float_compare
+
+
+class SalePaymentSheet(models.Model):
+    _name = "sale.payment.sheet"
+    _description = "Sale Payment Sheet"
+    _order = "date desc, id desc"
+    _inherit = ["mail.thread"]
+
```
Comment (rousseldenis) [inline sale_payment_sheet/models/sale_payment_sheet.py:16]: IMHO _check_company_auto = True is required here.

File: sale_payment_sheet/models/sale_payment_sheet.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+    )
+    statement_id = fields.Many2one(
+        comodel_name="account.bank.statement", string="Bank statement"
+    )
+
+    @api.model
+    def _default_journal(self):
+        return self.env.user.commercial_journal_ids[:1]
+
+    @api.depends("journal_id")
```
Comment (rousseldenis) [inline sale_payment_sheet/models/sale_payment_sheet.py]: Should be journal_id.currency_id

File: sale_payment_sheet/models/sale_payment_sheet.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+        for sheet in sheets:
+            statement = BankStatement.create(
+                {
+                    "name": sheet.name,
+                    "date": sheet.date,
+                    "journal_id": sheet.journal_id.id,
+                    "user_id": sheet.user_id.id,
+                }
+            )
+            for line in sheet.line_ids:
```
Comment (rousseldenis) [inline sale_payment_sheet/models/sale_payment_sheet.py:145]: Why not using the create multi ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Some comments. LGTM

File: sale_payment_sheet/models/sale_payment_sheet.py (hunk 6 above)
Comment (sergio-teruel) [reply sale_payment_sheet/models/sale_payment_sheet.py:145]: Because I am linking each statement line to the sheet line.

**Comment by sergio-teruel** [comment]
@rousseldenis Changes done!!

File: sale_payment_sheet/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+{
+    "name": "Sale payment sheet",
+    "summary": "Allow to create invoice payments to commercial users without "
+    "accounting permissions",
+    "version": "13.0.1.0.0",
+    "development_status": "Beta",
+    "category": "Account",
+    "website": "https://github.com/OCA/sale-workflow/",
+    "author": "Tecnativa, Odoo Community Association (OCA)",
```
Comment (victoralmau) [inline sale_payment_sheet/__manifest__.py:12]:
```suggestion
    "author": "Tecnativa, Odoo Community Association (OCA)",
    "maintainer": "sergio-teruel",
```

Comment (pedrobaeza) [reply sale_payment_sheet/__manifest__.py:12]:
```suggestion
    "author": "Tecnativa, Odoo Community Association (OCA)",
    "maintainers": ["sergio-teruel"],
```

File: sale_payment_sheet/models/sale_payment_sheet.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+    _name = "sale.payment.sheet.line"
+    _description = "Sale Payment Sheet Line"
+    _order = "sheet_id desc, date, sequence, id desc"
+
+    name = fields.Char(
+        string="Label", compute="_compute_name", store=True, readonly=False
+    )
+    date = fields.Date(
+        required=True,
+        default=lambda self: self._context.get("date", fields.Date.context_today(self)),
```
Comment (rousseldenis) [inline sale_payment_sheet/models/sale_payment_sheet.py]: Deprecated a long time ago. Please use self.env.context instead

https://github.com/odoo/odoo/blob/14.0/odoo/models.py#L4974

**Comment by sergio-teruel** [comment]
@rousseldenis Changes done!!! thanks...

**Pull request merged**